File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Zebpay.java
package com.mobnetic.coinguardian.model.market;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;

public class Zebpay extends Market {

    private final static String NAME = "Zebpay";
    private final static String TTS_NAME = "Zeb Pay";
    private final static String URL = "https://www.zebapi.com/api/v1/market/ticker-new/%1$s/%2$s";
    private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<>();

    static {
        CURRENCY_PAIRS.put(VirtualCurrency.AE, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.BAT, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.BCH, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
                VirtualCurrency.TUSD
        });
        CURRENCY_PAIRS.put(VirtualCurrency.BTG, new String[]{
                VirtualCurrency.BTC,
                VirtualCurrency.TUSD
        });
        CURRENCY_PAIRS.put(VirtualCurrency.CMT, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.EOS, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ETH, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.GNT, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.IOST, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.KNC, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.LTC, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.NCASH, new String[]{
                VirtualCurrency.BTC,
                VirtualCurrency.XRP
        });
        CURRENCY_PAIRS.put(VirtualCurrency.OMG, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.REP, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.TRX, new String[]{
                VirtualCurrency.BTC,
                VirtualCurrency.XRP
        });
        CURRENCY_PAIRS.put(VirtualCurrency.XRP, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ZIL, new String[]{
                VirtualCurrency.BTC
        });
        CURRENCY_PAIRS.put(VirtualCurrency.ZRX, new String[]{
                VirtualCurrency.BTC
        });
    }

    public Zebpay() {
        super(NAME, TTS_NAME, CURRENCY_PAIRS);
    }

    @Override
    public String getUrl(int requestId, CheckerInfo checkerInfo) {
        return String.format(URL, checkerInfo.getCurrencyBaseLowerCase(), checkerInfo.getCurrencyCounterLowerCase());
    }

    @Override
    protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
        ticker.bid = jsonObject.getDouble("sell");
        ticker.ask = jsonObject.getDouble("buy");
        ticker.vol = jsonObject.getDouble("volume");
        ticker.last = jsonObject.getDouble("market");
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Zebpay.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.currency.CurrencyPairsMap
import com.aneonex.bitcoinchecker.datamodule.model.currency.VirtualCurrency
import org.json.JSONObject

class Zebpay : Market(NAME, TTS_NAME, CURRENCY_PAIRS) {
    companion object {
        private const val NAME = "Zebpay"
        private const val TTS_NAME = "Zeb Pay"
        private const val URL = "https://www.zebapi.com/api/v1/market/ticker-new/%1\$s/%2\$s"
        private val CURRENCY_PAIRS: CurrencyPairsMap = CurrencyPairsMap()

        init {
            CURRENCY_PAIRS[VirtualCurrency.AE] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.BAT] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.BCH] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.BTC] = arrayOf(
                    VirtualCurrency.TUSD
            )
            CURRENCY_PAIRS[VirtualCurrency.BTG] = arrayOf(
                    VirtualCurrency.BTC,
                    VirtualCurrency.TUSD
            )
            CURRENCY_PAIRS[VirtualCurrency.CMT] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.EOS] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.ETH] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.GNT] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.IOST] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.KNC] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.LTC] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.NCASH] = arrayOf(
                    VirtualCurrency.BTC,
                    VirtualCurrency.XRP
            )
            CURRENCY_PAIRS[VirtualCurrency.OMG] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.REP] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.TRX] = arrayOf(
                    VirtualCurrency.BTC,
                    VirtualCurrency.XRP
            )
            CURRENCY_PAIRS[VirtualCurrency.XRP] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.ZIL] = arrayOf(
                    VirtualCurrency.BTC
            )
            CURRENCY_PAIRS[VirtualCurrency.ZRX] = arrayOf(
                    VirtualCurrency.BTC
            )
        }
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyBaseLowerCase, checkerInfo.currencyCounterLowerCase)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        ticker.bid = jsonObject.getDouble("sell")
        ticker.ask = jsonObject.getDouble("buy")
        ticker.vol = jsonObject.getDouble("volume")
        ticker.last = jsonObject.getDouble("market")
    }
}